source/common/http/header_utility.cc

```diff
@@ -259,6 +260,22 @@ HeaderUtility::requestHeadersValid(const RequestHeaderMap& headers) {
   return absl::nullopt;
 }
 
+absl::optional<std::pair<std::reference_wrapper<const absl::string_view>, Http::Code>>
+HeaderUtility::requestHeadersValidAtHcm(const RequestHeaderMap& headers,
+                                        uint32_t xff_num_trusted_hops, bool connection_is_ssl) {
+  if (xff_num_trusted_hops == 0 &&
```

Comment:
Can you add more comments about the xff_num_trusted_hops check? Every time I wade back into this code I get pretty confused and it would be good to leave some crumbs for the next person.

Reply:
Per some other comment threads, I'm pretty scared of breaking something with this new check. I'm almost definitely not fully understanding, but perhaps in the extra comments can you explain why this is safe and shouldn't break anyone today? Especially with how this check interacts with the fact that we still set upstream scheme?

Reply:
I had a giant 15 line comment going through all the cases, but let's sort out if we'd 
1) prefer to consistently validate non https-over-plaintext at the risk of breaking folks with previously working bad configs
2) not consistently validate, but make sure we validate the new case of using scheme of absolute URLs.
first :-)

Reply:
Can you spell out (1) more? I think I'm just not understanding the desired behavior and what we do today. I know for sure that people forward HTTPS incoming traffic to upstream HTTP, and I want to make sure we don't break that?

Reply:
I went back and refreshed my memory on all the xff settings, and I'm pretty sure we will break people if we don't also check that use remote address is true? Consider the case where an L1 forwards directly to an internal mesh over plaintext. Wouldn't that break unless we check if using remote address? Sorry for all the questions but I want to make sure I'm fully understanding this.

Reply:
so yeah, I could check use_remote and XFP.  Assuming we do that a way we're happy with if you look at the grid of 
1) edge envoys
2) properly configured non-edge (trust XFP/XFF)
3) badly configured non-edge (don't trust XFP/XFF)
this PR as is would validate properly for 1, be a no-op for 2, and break 3 for https-in-plaintext.

We can instead move the checks to the codec, at which point we'd validate https://absolute/url-over-plaintext at edge Envoys, allow https schemed http2 in plaintext at edge envoys (not a case I'm terribly worried about) and allow the badly configured envoys (3) to continue to be badly configured without breaking them.

Reply:
I think we are talking past each other, or I'm very confused (quite likely). Let me be really specific with the case I am concerned about:
1) Edge Envoy configured with use_remote_address = true, num_trusted_hop = 0, terminating TLS. In this case we don't trust XFF/XFP, set them, and with this change also either set scheme = XFP or verify it is not HTTPS over plaintext. So far so good. This makes sense.
2) Envoy 1 forwards to a mesh Envoy over plain text (either just because, or because using IPSEC, or whatever - I think this is a perfectly valid use case). Mesh Envoy is configured with use_remote_address = false, num_trusted_hop = 0. Therefore it gets the address from XFF and respects XFP. Scheme in this case would be "https" in the common case.

With this change, wouldn't we now fail these requests which I think are legitimate? This is why I think this check needs to be gated on use_remote_address = true? Please tell me what I'm missing here. :)

Reply:
I agree we are talking past each other :-) 
I agree moving to use_remote_address makes sense and I'm happy to do that.  
I'm also trying to determine if we want to lower even that risk by moving these checks to the HTTP/1 codec, since the absolute URL https://-over-http use case is the only security risk opened by this pr.  I'll switch to use_remote once I know if I'm doing an in-place switch or moving-to-codec switch!

Reply:
Yeah +1 to just doing the most limited check possible. I think we could consider a more strict check later once the rest of the work is done?